Apim 4341 - List subscriptions

jansobczynski opened the pull request:
## Issue

https://gravitee.atlassian.net/browse/APIM-4341

## Description

Subscription list

## Additional context
Screenshot:
<img width="1109" alt="Screenshot 2024-05-21 at 2 50 41 PM" src="https://github.com/gravitee-io/gravitee-api-management/assets/144100648/dad02af9-7053-49ff-b77b-f02a5d24e090">
<!-- Storybook placeholder -->
---

📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-nxzauxqazc.chromatic.com)
<!-- Storybook placeholder end -->

jansobczynski committed:
feat(portal-next): list subscription

jourdiw commented on gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.html:
```diff
@@ -22,6 +22,29 @@
+
+      <ng-container matColumnDef="plan">
+        <th mat-header-cell *matHeaderCellDef i18n="@@subscriptionTableColumnPlan">Plan</th>
+        <td mat-cell *matCellDef="let element">{{ subscriptionsPlanName(element.plan) }}</td>
```
Plan name should be titlecase I think. To be verified with Anthony + Audrey.

jourdiw commented on gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.ts:
```diff
@@ -41,9 +45,18 @@ export class ApiTabSubscriptionsComponent implements OnInit {
   }
 
+  subscriptionsPlanName(plan: string) {
+    if (Object.hasOwn(this.subscriptionsMetadataList, plan)) {
```
Every subscription will have a plan associated and every plan must have a name. So if the metadata is unreliable (aka you have to include a solution for if the plan name), you should get a list of plans and then match the plan id used in the subscription with the appropriate plan to get the plan name.

jansobczynski replied:
I confirmed that all necessary data is in metadata object

jourdiw commented on gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.html:
```diff
@@ -22,6 +22,29 @@
+
+      <ng-container matColumnDef="status">
+        <th mat-header-cell *matHeaderCellDef i18n="@@subscriptionTableColumnStatus">Status</th>
+        <td mat-cell *matCellDef="let element">{{ element.status }}</td>
```
This should be title case as well

jansobczynski replied:
Done

jourdiw commented on gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.scss:
```diff
@@ -20,3 +20,17 @@
+  }
+
+  th.mat-mdc-header-cell {
+    font-weight: 600;
```
Use one of the classes in `m3-adapter` to change the font style.

jansobczynski replied:
Done

jourdiw commented on gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.scss:
```diff
@@ -20,3 +20,17 @@
+
+.api-tab-subscriptions__table {
+  overflow: hidden;
+  border: 1px solid #e6e0ec;
```
Hex codes shouldn't be used directly... reference the palette with a `color-mix`.

jansobczynski replied:
Done

jourdiw commented on gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.scss:
```diff
@@ -20,3 +20,17 @@
+  border: 1px solid #e6e0ec;
+  border-radius: 16px;
+
+  tr.mat-mdc-row.mdc-data-table__row {
```
Avoid using classes specific to material design. They can change names which will create style regressions in the future.

jansobczynski replied:
Done

jourdiw reviewed (commented):
Should add tests as well that the correct information is displayed.

jansobczynski commented:
Tests added

jourdiw commented on gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.html:
```diff
@@ -22,6 +22,29 @@
+    <table mat-table [dataSource]="subscriptions" class="api-tab-subscriptions__table">
+      <ng-container matColumnDef="application">
+        <th mat-header-cell *matHeaderCellDef i18n="@@subscriptionTableColumnApplication" class="m3-title-medium">Application</th>
+        <td mat-cell *matCellDef="let element">{{ retrieveMetadataName(element.application) }}</td>
```
Methods should be avoided in the html rendering because they are called incessantly, and in this case, the value will not be changing once rendered. Instead of doing the mapping here, move the mapping to the component after the service returns the Subscription response.

jansobczynski replied:
Done

jourdiw commented on gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.scss:
```diff
@@ -13,10 +13,23 @@
+  border-radius: 16px;
+
+  &-row {
+    background: #fff;
```
Instead of using hex-code, use a variable. You can reference the mat card background color so that it matches. If we ever change to dark-mode, it will follow the same choice as the mat-cards.

jansobczynski replied:
Done

jourdiw commented on gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.spec.ts:
```diff
@@ -47,6 +52,24 @@ describe('ApiTabSubscriptionsComponent', () => {
+    it('should show subscription list', async () => {
+      const subscriptionTable = await harnessLoader.getHarness(MatTableHarness.with({ selector: '.api-tab-subscriptions__table' }));
+      expect(subscriptionTable).toBeTruthy();
+      console.log(3, subscriptionTable);
```
Remove console.log

jansobczynski replied:
Done

jourdiw commented on gravitee-apim-portal-webui-next/src/entities/subscription/subscription.fixture.ts:
```diff
@@ -50,6 +50,11 @@ export function fakeSubscriptionResponse(
     links: {
       self: 'test',
     },
+    metadata: {
```
Since the application id should always exist with the name, it should be added here to reflect in the tests.

jansobczynski replied:
Done

jourdiw commented on gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.ts:
```diff
@@ -33,17 +37,28 @@ export class ApiTabSubscriptionsComponent implements OnInit {
     this.subscriptionsList$ = this.loadSubscriptions$();
   }
 
+  retrieveMetadataName(plan: string) {
```
It's a bit confusing that the parameter is named `plan` but you are using it for applications as well. Why not rename it to `id`?

jansobczynski replied:
Done

jourdiw commented on gravitee-apim-portal-webui-next/src/pipe/capitalize-first.pipe.ts:
```diff
@@ -0,0 +1,27 @@
+  standalone: true,
+})
+export class CapitalizeFirstPipe implements PipeTransform {
+  transform(value: string): string {
```
Maybe add a test that this works for empty strings `''` and one-letter strings, aka `'a'`?

jansobczynski replied:
Done

jourdiw reviewed (changes requested):
When all the changes are done, update the screenshot in the PR description to represent what you have developed. 👍

jansobczynski commented:
Screenshot update:
<img width="1061" alt="Screenshot 2024-05-24 at 11 16 39 AM" src="https://github.com/gravitee-io/gravitee-api-management/assets/144100648/ad9e0e97-6f8e-4ca4-ba6c-44e9cd767718">
I added one extra ui change, moved arrows to the right to match figma design
